### PR TITLE
Add PerAPILogging feature

### PR DESCRIPTION
**Pull request by binodmx**
Fixes https://github.com/wso2/product-apim/issues/11720

### CODE REVIEW

File: features/apimgt/org.wso2.carbon.apimgt.rest.api.devops.feature/pom.xml [hunk 1]
```diff
@@ -0,0 +1,165 @@
...
+                            <id>org.wso2.carbon.apimgt.rest.api.devops</id>
+                            <propertiesFile>../feature.properties</propertiesFile>
+                            <adviceFile>
+                                <properties>
+                                    <propertyDef>org.wso2.carbon.p2.category.type:server</propertyDef>
+                                    <propertyDef>org.eclipse.equinox.p2.type.group:false</propertyDef>
+                                </properties>
+                            </adviceFile>
+                            <bundles>
+                                <bundleDef>org.wso2.carbon.identity.framework:org.wso2.carbon.identity.entitlement.stub:${carbon.identity.version}</bundleDef>
```
Comment (tharindu1st) [inline features/apimgt/org.wso2.carbon.apimgt.rest.api.devops.feature/pom.xml:153]: is there any reason to need this

File: components/apimgt/org.wso2.carbon.apimgt.rest.api.devops/src/main/webapp/WEB-INF/web.xml [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+        <param-name>contextConfigLocation</param-name>
+        <param-value>WEB-INF/beans.xml</param-value>
+    </context-param>
+    <listener>
+        <listener-class>
+            org.springframework.web.context.ContextLoaderListener
+        </listener-class>
+    </listener>
+    <servlet>
+        <description>JAX-WS/JAX-RS Device Registration Agent Endpoint</description>
```
Comment (tharindu1st) [inline components/apimgt/org.wso2.carbon.apimgt.rest.api.devops/src/main/webapp/WEB-INF/web.xml:39]: change the description

**Review by tharindu1st — CHANGES_REQUESTED** [review]

File: components/apimgt/org.wso2.carbon.apimgt.devops.impl/pom.xml [hunk 3]
```diff
@@ -0,0 +1,104 @@
...
+                        <Bundle-Name>${project.artifactId}</Bundle-Name>
+                        <Export-Package>
+                            org.wso2.carbon.apimgt.devops.impl.*;version="${carbon.apimgt.version}"
+                        </Export-Package>
+                        <Import-Package>
+                            org.wso2.carbon.apimgt.api.*;version="${carbon.apimgt.imp.pkg.version}",
+                            org.wso2.carbon.apimgt.gateway.*;version="${carbon.apimgt.imp.pkg.version}",
+                            org.wso2.carbon.apimgt.impl.*;version="${carbon.apimgt.imp.pkg.version}",
+                        </Import-Package>
+                        <DynamicImport-Package>*</DynamicImport-Package>
```
[the remark this answers is not in the document]
Comment (binodmx) [reply components/apimgt/org.wso2.carbon.apimgt.devops.impl/pom.xml]: removed

File: components/apimgt/org.wso2.carbon.apimgt.devops.impl/pom.xml [hunk 4]
```diff
@@ -0,0 +1,104 @@
+<?xml version="1.0" encoding="utf-8"?>
+<!--
+  ~ Copyright (c) 2021, WSO2 Inc. (http://www.wso2.org) All Rights Reserved.
```
[the remark this answers is not in the document]
Comment (binodmx) [reply components/apimgt/org.wso2.carbon.apimgt.devops.impl/pom.xml]: changed year to 2022 in each **newly added** file

File: components/apimgt/org.wso2.carbon.apimgt.rest.api.devops/src/main/webapp/WEB-INF/web.xml (hunk 2 above)
Comment (binodmx) [reply components/apimgt/org.wso2.carbon.apimgt.rest.api.devops/src/main/webapp/WEB-INF/web.xml]: changed the display name and description in line 27-28 and removed this line adhering to web.xml files in publisher/store/gateway rest apis.

File: components/apimgt/org.wso2.carbon.apimgt.devops.impl/src/main/java/org.wso2.carbon.apimgt.devops.impl/logging/PerAPILoggingImpl.java [hunk 5]
```diff
@@ -0,0 +1,73 @@
...
+        if (!APIUtil.hasPermission(RestApiCommonUtil.getLoggedInUsername(), PER_API_LOGGING_PERMISSION_PATH)) {
+            throw new APIManagementException("Invalid logging permission",
+                    ExceptionCodes.from(ExceptionCodes.INVALID_PERMISSION));
+        }
+        LoggingMgtDAO.getInstance().addAPILogger(tenantId, apiId, logLevel);
+        publishLogAPIData(tenantId, apiId, logLevel);
+    }
+
+    private void publishLogAPIData(String tenant, String apiId, String value) throws APIManagementException {
+        APIProvider apiProvider = RestApiCommonUtil.getLoggedInUserProvider();
```
[the remark this answers is not in the document]
Comment (binodmx) [reply components/apimgt/org.wso2.carbon.apimgt.devops.impl/src/main/java/org.wso2.carbon.apimgt.devops.impl/logging/PerAPILoggingImpl.java]: used `ApiMgtDAO.getInstance().getAPIContext(apiId)` to get the APIContext and removed apiProvider.

File: features/apimgt/org.wso2.carbon.apimgt.rest.api.devops.feature/pom.xml (hunk 1 above)
Comment (binodmx) [reply features/apimgt/org.wso2.carbon.apimgt.rest.api.devops.feature/pom.xml:153]: keeping this line to adhere with other rest api features.

**Comment by bhathiya** [comment]
Do we have a corresponding integration tests PR?

**Review by tharindu1st — APPROVED** [review]

**Pull request merged**